### DISCUSSION TITLE
error connecting to api.github.com

### DISCUSSION THREAD

**Discussion by benkasminbullock** [community/community#134707] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I'm unable to connect to api.github.com with gh. I am using the FreeBSD version of gh which I installed using `pkg install`. The error which occurs is `error connecting to api.github.com`. However using Telnet it is clear that the computer is able to reach the host on the HTTPS port 443. Please see the screenshot below.

![github](https://github.com/user-attachments/assets/f6f7c3c6-11b7-4740-9116-3d500537930d)

**Comment by officialprosingh** [comment]
Here are a few steps you can try to resolve this issue:

Verify the Authentication Token:

Ensure that the token you're using has the correct scopes (repo, read:org, workflow).
Double-check that the token has not expired or been revoked. You can generate a new token from [GitHub's Personal Access Tokens page](https://github.com/settings/tokens).
Check Network Configuration:

Although Telnet confirms that you can reach api.github.com on port 443, there might be firewall rules or proxy settings that could be interfering with the GitHub CLI.
Try running the following command to check if you can reach api.github.com using curl:
bash
Copy code
curl -v https://api.github.com
If there are any issues with curl, it might give you more detailed information about what is going wrong.
Update GitHub CLI:

Make sure you have the latest version of the GitHub CLI installed. You can update it using pkg:
bash
Copy code
sudo pkg update
sudo pkg upgrade gh
Check Environment Variables:

Ensure that your environment variables are correctly set for the GitHub CLI to use the token. For example, you can set the GH_TOKEN variable:

**Reply by benkasminbullock** [reply]
> Here are a few steps you can try to resolve this issue:

Thank you for this careful response. I appreciate your taking the time to write this. Unfortunately it did not resolve the problem.
 
> Verify the Authentication Token:
> 
> Ensure that the token you're using has the correct scopes (repo, read:org, workflow). Double-check that the token has not expired or been revoked. You can generate a new token from [GitHub's Personal Access Tokens page](https://github.com/settings/tokens). Check Network Configuration:

I checked `admin:org` and `workflow` which seems to automatically give me `repo`. The tokens were created yesterday immediately after starting `gh` and had not expired.

> Although Telnet confirms that you can reach api.github.com on port 443, there might be firewall rules or proxy settings that could be interfering with the GitHub CLI. Try running the following command to check if you can reach api.github.com using curl: bash Copy code curl -v https://api.github.com If there are any issues with curl, it might give you more detailed information about what is going wrong.

```
curl -v https://api.github.com
* Host api.github.com:443 was resolved.
* IPv6: (none)
* IPv4: 20.27.177.116
*   Trying 20.27.177.116:443...
* Connected to api.github.com (20.27.177.116) port 443
```
There are many more lines, but the connection was clearly successful.

> Update GitHub CLI:
> 
> Make sure you have the latest version of the GitHub CLI installed. You can update it using pkg: bash Copy code sudo pkg update sudo pkg upgrade gh Check Environment Variables:
>

````
# pkg upgrade gh
Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
Checking integrity... done (0 conflicting)
Your packages are up to date.
````

![github 20240806](https://github.com/user-attachments/assets/5c45392e-2fa3-43ff-8462-db5c696626ab)

 
> Ensure that your environment variables are correctly set for the GitHub CLI to use the token. For example, you can set the GH_TOKEN variable:

After the above screenshot, I set `GH_TOKEN` to the exact value given on the web browser using `export GH_TOKEN=*****` where `*****` represents the value copy-pasted from https://github.com/settings/tokens. I got the following sort of output:

```
[ben@mikan] {06:32 08} ~ 512 $ gh auth  status
github.com
  X Failed to log in to github.com using token (GH_TOKEN)
  - Active account: true
  - The token in GH_TOKEN is invalid.
[ben@mikan] {06:32 38} ~ 513 $ gh auth login
? What account do you want to log into? GitHub.com
The value of the GH_TOKEN environment variable is being used for authentication.
To have GitHub CLI store credentials instead, first clear the value from the environment.
```
I don't know if it is relevant but the value is copy-pasted from Windows into the FreeBSD terminal screen running under XWindows.

**Reply by ebndev** [reply]
Hi @officialprosingh, thanks for participating in GitHub Discussions!

Please note: we’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  

You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. 

Please feel encouraged to continue engaging in Discussions as your authentic self though. We appreciate you being a part of the Community 🌟

**Comment by ebndev** [comment]
Hi @benkasminbullock, thanks for posting in GitHub Discussions! 

I've gone ahead and moved this post to our `API & Webhooks` category, as this topic may be more relevant there and might help someone give you a nudge in the right direction. 😁

In the meantime, please feel encouraged to explore existing discussions in the category and utilize the search in case other users may have run into the same issue!

**Reply by benkasminbullock** [reply]
> Hi @benkasminbullock, thanks for posting in GitHub Discussions!

There is really no need to thank me. I want to use `gh` to add files to a release on github but this problem has made it impossible.

> I've gone ahead and moved this post to our `API & Webhooks` category, as this topic may be more relevant there and might help someone give you a nudge in the right direction. 😁

Are you sure that is the correct place? This question is about the `gh` command line tool but the API and Webhooks seems to not be about that. Did you just look at the letters api in my question and decide to move it here, or is there a better reason?

> In the meantime, please feel encouraged to explore existing discussions in the category and utilize the search in case other users may have run into the same issue!

I have utilised search to try to find other users with the same issue but have not located any. In fact my query is coming out at the top of search results now:

https://www.google.com/search?hl=en&q=%22error%20connecting%20to%20api.github.com%22

The ungrammatical error message is not very encouraging either.

**Reply by SunShineHead** [reply]
```

```

**Comment by benkasminbullock** [comment]
The error seems to originate from https://github.com/cli/cli/blob/86964d8809bed7bfbd798d444d5923b8a2bbd4d6/cmd/gh/main.go#L205.

I had a look with tcpdump and this is the only interaction I saw:

````
16:02:52.155375 IP mikan.42925 > rtx1000.domain: 41282+ [1au] A? api.github.com. (43)
16:02:52.155380 IP mikan.32037 > rtx1000.domain: 61222+ [1au] AAAA? api.github.com. (43)
16:02:52.155935 IP rtx1000.domain > mikan.42925: 41282 FormErr- 0/0/0 (32)
16:02:52.155943 IP rtx1000.domain > mikan.32037: 61222 FormErr- 0/0/0 (32)
````

The router (rtx1000) seems to be returning something called `FormErr` after the initial DNS query which causes the crash.

**Reply by benkasminbullock** [reply]
I downloaded the source code of cli but there was some sort of odd problem where it didn't want to build. Looking at the source code I found that it's possible to switch on debugging by setting an environment variable `export GH_DEBUG=1`. So I tried switching it on and the output of the error looks like this now:
```
* Request at 2024-08-07 16:32:17.964095788 +0900 JST m=+561.184521704
* Request to https://api.github.com/
* dial tcp: lookup api.github.com on 192.168.2.1:53: server misbehaving
* Request took 97.257799ms
error connecting to api.github.com
lookup api.github.com on 192.168.2.1:53: server misbehaving
check your internet connection or https://githubstatus.com
``` 
192.168.2.1 is the address of the router which also has the name rtx1000 (it is actually an RTX 1200 but I forgot to change the name). The error `server misbehaving` comes from here:

https://github.com/golang/go/blob/e705a2d16e4ece77e08e80c168382cdb02890f5b/src/net/dnsclient_unix.go#L249

It is not very clear what sort of misbehaviour the router is engaging in. I note that every single internet access in my household has been going through this router for about ten years with absolutely no problem until now. Or perhaps it is the upstream server.

**Reply by benkasminbullock** [reply]
I tried a simple program to test this:
```
package main

import (
        "bufio"
        "fmt"
        "net"
)

func main() {
        conn, err := net.Dial("tcp", "www.google.com:443")
        //      conn, err := net.Dial("tcp", "api.github.com:443")
        if err != nil {
                fmt.Printf("Error: %s\n", err)
                return
        }
        fmt.Fprintf(conn, "GET / HTTP/1.0\r\n\r\n")
        status, err := bufio.NewReader(conn).ReadString('\n')
        if err != nil {
                fmt.Printf("Error: %s\n", err)
                return
        }
        fmt.Printf("%v\n", status)
}
```
It gives `server misbehaving` either with www.google.com or api.github.com. I tried updating the firmware on the router to see if that would help but it didn't make any difference. It's not at all clear what the message "server misbehaving" even means, as pointed out here:
https://github.com/golang/go/issues/44208

**Reply by benkasminbullock** [reply, marked as answer]
I installed some tools called `bind-tools` to investigate this situation. Looking at the output of `dig api.github.com A` it turned out that the router-based nameserver did not understand `edns` (extended DNS) at all, and was sending the `FORMERR` response.

The router I am using, Yamaha RTX1200, doesn't appear to offer an EDNS facility.  I don't know if there is a way to get the Go net package to not insist on `edns`. and rather than spend a long time trying to solve that problem, what I did was to edit `/etc/resolv.conf` to add a line `8.8.8.8` to add the Google DNS service. After doing that, I was able to log in using `gh`.

![github 20240807](https://github.com/user-attachments/assets/079da8f2-49cb-47d9-8ea2-2c70d4d2329a)

So the problem was due to an antiquated router which was incompatible with Go's net package and the solution was to use the Google nameserver instead of going via the router.

Thank you for your assistance.

**Reply by nobin12-ag** [reply]
Troubleshoot

**Reply by XavierMP14** [reply]
X

**Comment by barnygon** [comment]
Delete all comsola

**Comment by yaswanthteja** [comment]
### 🛠️ Problem
- `gh auth login` fails with:  
  `error connecting to api.github.com`

### ✅ Confirmed
- Internet works (Telnet connected to api.github.com ✅)

### 🔧 Fix

1. Check for a **corporate firewall/proxy** blocking `gh` CLI:
   - Try: `gh auth login --hostname github.com --insecure` (⚠️ temporary workaround)
   - Or configure proxy:  
     ```bash
     export HTTPS_PROXY=http://your.proxy:port
     ```

2. Ensure latest GitHub CLI:
   ```bash
   gh version
   gh upgrade
```
Try using SSH instead of HTTPS if issue persists:

```
gh auth login
```
# choose SSH as protocol

**Reply by tacsizkral546154** [reply]
Ok

**Comment by bossbdpro1** [comment]
@tag:workspaceTrust
